src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.KeyServices.cs

```diff
@@ -0,0 +1,149 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using Microsoft.Extensions.DependencyInjection;
+namespace Microsoft.AspNetCore.Http.Generators.Tests;
+
+public partial class RequestDelegateCreationTests : RequestDelegateCreationTestBase
+{
+    [Fact]
+    public async Task SupportsSingleKeyedServiceWithStringKey()
+    {
+        var source = """
+app.MapGet("/", (HttpContext context, [FromKeyedServices("service1")] TestService arg) => context.Items["arg"] = arg);
+""";
+        var (_, compilation) = await RunGeneratorAsync(source);
+        var myOriginalService = new TestService();
+        var serviceProvider = CreateServiceProvider((serviceCollection) => serviceCollection.AddKeyedSingleton("service1", myOriginalService));
+        var endpoint = GetEndpointFromCompilation(compilation, serviceProvider: serviceProvider);
+
+        var httpContext = CreateHttpContext(serviceProvider);
+        await endpoint.RequestDelegate(httpContext);
+
+        Assert.Same(myOriginalService, httpContext.Items["arg"]);
+    }
+
+    [Fact]
+    public async Task SupportsSingleKeyedServiceWithCharKey()
+    {
+        var source = """
+app.MapGet("/", (HttpContext context, [FromKeyedServices('a')] TestService arg) => context.Items["arg"] = arg);
+""";
+        var (_, compilation) = await RunGeneratorAsync(source);
+        var myOriginalService = new TestService();
+        var serviceProvider = CreateServiceProvider((serviceCollection) => serviceCollection.AddKeyedSingleton('a', myOriginalService));
+        var endpoint = GetEndpointFromCompilation(compilation, serviceProvider: serviceProvider);
+
+        var httpContext = CreateHttpContext(serviceProvider);
+        await endpoint.RequestDelegate(httpContext);
+
+        Assert.Same(myOriginalService, httpContext.Items["arg"]);
+    }
+
+    [Theory]
+    [InlineData(10)]
+    [InlineData(false)]
+    [InlineData(12.3)]
+    public async Task SupportsSingleKeyedServiceWithPrimitiveKeyTypes(object key)
+    {
+        var source = $$"""
+app.MapGet("/", (HttpContext context, [FromKeyedServices({{key.ToString()?.ToLowerInvariant()}})] TestService arg) => context.Items["arg"] = arg);
```

Comment:
Does this work? The key is a `string` but the keyed service is added with `object`.

Reply:
The key here is actually an object. The `ToString()?.ToLowerVariant()` here is to get it to emit the write code for the `false` value. The source code snippet ends up having `[FromKeyedServices(false)]`, `[FromKeyedServices(12.3)]`, etc.

Reply:
Ok, stringed code is odd looking 😃